pkg/segment/writer/agiletree.go

```diff
@@ -222,30 +216,64 @@ func (stb *StarTreeBuilder) DropSegTree(stbDictEncWorkBuf [][]string) {
 	stb.ResetSegTree(stb.groupByKeys, stb.mColNames, stbDictEncWorkBuf)
 }
 
+func (stb *StarTreeBuilder) DropColumns(colsToDrop []string) error {
+	if len(colsToDrop) == 0 {
+		return nil
+	}
+
+	mapColNameToIdx := make(map[string]int)
+	dropIndexes := make(map[int]struct{})
+
+	// check if all columns to drop are present
+	for idx, colName := range stb.groupByKeys {
+		mapColNameToIdx[colName] = idx
+	}
+
+	// check if all columns to drop are present
+	for _, colToDrop := range colsToDrop {
+		colIdx, exists := mapColNameToIdx[colToDrop]
+		if !exists {
+			return fmt.Errorf("DropColumns: column to drop %v not found", colToDrop)
+		}
+		err := stb.DropColumn(colToDrop)
+		if err != nil {
+			return fmt.Errorf("DropColumns: Error while dropping column %v, err: %v", colToDrop, err)
+		}
+		dropIndexes[colIdx] = struct{}{}
+	}
+
+	stb.segDictMap = toputils.RemoveElements(stb.segDictMap, dropIndexes)
+	stb.segDictEncRev = toputils.RemoveElements(stb.segDictEncRev, dropIndexes)
+	stb.segDictLastNum = toputils.RemoveElements(stb.segDictLastNum, dropIndexes)
+
+	// No need to update wipRecNumToColEnc, since it will be reset in each ComputeStarTree based on groupByKeys
+	stb.wipRecNumToColEnc = stb.wipRecNumToColEnc[:stb.numGroupByCols]
+
+	return nil
+}
+
 func (stb *StarTreeBuilder) DropColumn(colToDrop string) error {
+	newGrpByKeys := make([]string, 0, len(stb.groupByKeys)-1)
 	dropLevel := -1
 	for idx, col := range stb.groupByKeys {
 		if col == colToDrop {
 			dropLevel = idx
 			continue
 		}
+		newGrpByKeys = append(newGrpByKeys, col)
 	}
+
 	if dropLevel == -1 {
 		return fmt.Errorf("DropColumn: column to drop %v not found", colToDrop)
 	}
-	if stb.droppedCols[dropLevel] {
-		return fmt.Errorf("DropColumn: column %v already dropped", colToDrop)
-	}
-	nextLevel := stb.getNextLevel(-1)
-	if nextLevel == -1 {
-		return fmt.Errorf("DropColumn: cannot drop last column")
-	}
-	err := stb.removeLevelFromTree(stb.tree.Root, uint(nextLevel), uint(dropLevel))
+	err := stb.removeLevelFromTree(stb.tree.Root, 0, uint(dropLevel))
 	if err != nil {
 		return err
 	}
+	stb.numGroupByCols--
+	stb.groupByKeys = newGrpByKeys
 
-	stb.droppedCols[dropLevel] = true
+	// Rest of the cleanup is done in DropColumns for efficiency
 
 	return nil
```

Comment:
You're inside `DropColumn()` which is a public function, so you can't rely on `DropColumns()` being the the caller. Either make `DropColumn()` private, or do the cleanup here

Reply:
addressed in the latest commit.